### ISSUE TITLE
chrome$Event intersections result in the wrong types

### ISSUE THREAD

**Issue by callumlocke** [Shraymonks/flow-interfaces-chrome#23]
Flow tells me the `port` argument here might be `void`:

![image](https://user-images.githubusercontent.com/250617/45591340-65b9d580-b950-11e8-9868-548d69dc357c.png)

But this is not true. It is always a defined port object. I think the problem is this `onConnect` event (like many others) is typed as an intersection with `chrome$Event`. This provides a competing definition for the `addListener` property that is wrong because it says the callback takes no arguments. I think the intention is probably that the latter part of the intersection should should take precedence (in this case the one that says the `addListener` callback takes a `chrome$Port` argument) but this isn't how Flow works.

**Comment by ssorallen** [comment]
I suspected this was the root cause of a problem I was facing. Thanks for opening the issue. 

The current types are trying to “override” the `addListener` definition, but intersecting two Objects in Flow actually intersects each of the two Objects’ properties as well, which is why you are getting two possible signatures for the place you are using this. Is it possible to do an Interface / Implementation setup where you can properly overwrite a signature rather than intersect it?

**Comment by ssorallen** [comment]
Intersecting Objects in Flow also [intersects properties of the same name](https://flow.org/en/docs/types/intersections/#toc-intersections-of-object-types); there is no "overriding" like what the current types in this repo are trying to do.

> But when you have properties that overlap by having the same name, it creates an intersection of the property type as well.

One way to fix this would be to have an `type chrome$IEvent` that does not include an `addListener` property and make `type chrome$Event` an intersection. All places that have a different signature for `addListener` would intersect with `chrome$IEvent`:

```
type chrome$IEvent {
  removeListener(callback: () => void): void,
  hasListener(callback: () => void): boolean,
  hasListeners(): boolean,

  addRules(rules: Array<chrome$Rule>, callback?: (rules: Array<chrome$Rule>) => void): void,
  getRules: (
    ((ruleIdentifiers: Array<string>, callback: (rules: Array<chrome$Rule>) => void) => void) &
    ((callback: (rules: Array<chrome$Rule>) => void) => void)
  ),
  removeRules(ruleIdentifiers?: Array<string>, callback?: () => void): void
}

type chrome$Event = chrome$IEvent & {
  addListener(callback: () => void): void,
}
```

then in runtime.js, for example:

```
type chrome$runtime = {
  ...
  onConnect: chrome$IEvent & {
    addListener(callback: (port: chrome$Port) => void): void
  },
  ...
}
```

**Comment by Shraymonks** [comment]
Do you have a minimal repro fiddle for the error using intersections? I can't seem to reproduce in [my testing](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVAXAngBwKZgDGAFgE5wC2eAJAKIBueAdhmALxgDeqYYAhgBMBAGQCWAZwzM8pABSE+MGACM+hANYAuMLICU7AHxh6cUQN3aTZgDSoAvgG502fETKUaABTilWHbrwCEoRwTEx4hBjaAGIArkyRoqG2js64BCTkVNQASvEYolTsXDy8YKEAwqHhkdqZHnSMLGAAZCVl-EJiktJyCkqqGtqyOD5RblleY-psRlbmlqYCpXb2TqgCETB8pAT0OxMe2gFgpPmFeHXu2XksFynr9VQAdGd3L5XVERjPgiISUnCchG00M7TAo18zyC4hCYW+eicdl0DiAA).

I think the best way to fix this is to use [object type spread](https://github.com/facebook/flow/issues/1326) instead of a type intersection. This will give the "overwrite" behavior that you both are describing; I used intersections originally since type spread wasn't yet implemented.

Here's an example of what `onConnect` would look like
```javascript
type chrome$runtime = {
  ...
  onConnect: {
      ...chrome$Event,
      addListener(callback: (port: chrome$Port) => void): void
    },
  ...
}
```

**Comment by ssorallen** [comment]
@ShrayMonks I didn’t realize you could use the spread operator in type definitions, that’s great. I’ll get a basic case that reproduces the issue.

**Comment by ssorallen** [comment]
If I put `manager` inside index.js, the Flow error does not happen. There's something happening with the exporting/importing. Separate modules always raises this error though.

**index.js**:
```
/* @flow */

import manager from './manager';

chrome.runtime.onConnect.addListener(manager.onConnect);
```

**manager.js**
```
/* @flow */

const manager = {
  onConnect(port: chrome$Port) {
  },
}

export default manager;
```

Flow output:
```
Error ┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ index.js:5:38

function [1] requires another argument from function type [2].

     index.js
      2│
      3│ import manager from './manager';
      4│
      5│ chrome.runtime.onConnect.addListener(manager.onConnect);
      6│

     node_modules/flow-interfaces-chrome/interfaces/events.js
 [2] 10│   addListener(callback: () => void): void,

     manager.js
 [1]  4│   onConnect(port: chrome$Port) {
```

**Comment by ssorallen** [comment]
I think this can be closed thanks to #24.
